j5/backends/hardware/j5/arduino.py

```diff
@@ -0,0 +1,228 @@
+
+from abc import abstractmethod
+from datetime import timedelta
+from threading import Lock
+from typing import Callable, List, Mapping, Optional, Set, Type
+
+from serial import Serial
+from serial.tools.list_ports import comports
+from serial.tools.list_ports_common import ListPortInfo
+
+from j5.backends.hardware.env import NotSupportedByHardwareError
+from j5.backends.hardware.j5.serial import SerialHardwareBackend
+from j5.boards import Board
+from j5.boards.arduino import ArduinoUno
+from j5.components import GPIOPinInterface, GPIOPinMode, LEDInterface
+
+
+class DigitalPinData:
+    """Contains data about a digital pin."""
+
+    mode: GPIOPinMode
+    state: bool
+
+    def __init__(self, *, mode: GPIOPinMode, state: bool):
+        self.mode = mode
+        self.state = state
+
+
+class ArduinoHardwareBackend(
+    LEDInterface,
+    GPIOPinInterface,
+    SerialHardwareBackend,
+):
+    """An abstract class to create backends for different Arduinos."""
+
+    board: Type[ArduinoUno]
+    USB_IDS: Set[Tuple[int, int]] = {
+        (0x2341, 0x0043),  # Fake Uno
+        (0x2a03, 0x0043),  # Fake Uno
+        (0x1a86, 0x7523),  # Real Uno
+        (0x10c4, 0xea60),  # Ruggeduino
+    }
+    DEFAULT_TIMEOUT: timedelta = timedelta(milliseconds=1250)
+
+    @classmethod
+    def is_arduino(port: ListPortInfo) -> bool:
+        """Check if a ListPortInfo represents a valid Arduino derivative."""
+        return (port.vid, port.pid) in USB_IDS
+
+    @classmethod
+    def discover(
+            cls,
+            comports: Callable = comports,
+            serial_class: Type[Serial] = Serial,
+    ) -> Set[Board]:
+        """Discover all connected arduino boards."""
+        # Find all serial ports.
+        ports: List[ListPortInfo] = comports()
+
+        # Get a list of boards from the ports.
+        boards: Set[Board] = set()
+        for port in filter(cls.is_arduino, ports):
+            boards.add(
+                cls.board(
+                    port.serial_number,
+                    cls(port.device, serial_class),
+                ),
+            )
+
+        return boards
+
+    def __init__(
+            self,
+            serial_port: str,
+            serial_class: Type[Serial] = Serial,
+            baud: int = 115200,
+            timeout: timedelta = DEFAULT_TIMEOUT,
+    ) -> None:
+        super(ArduinoHardwareBackend, self).__init__(
+            serial_port=serial_port,
+            serial_class=serial_class,
+            baud=baud,
+            timeout=timeout,
+        )
+
+        self.serial_port = serial_port
+
+        self._lock = Lock()
+
+        self._digital_pins: Mapping[int, DigitalPinData] = {
+            i: DigitalPinData(mode=GPIOPinMode.DIGITAL_INPUT, state=False)
+            for i in range(2, ArduinoUno.FIRST_ANALOGUE_PIN)
+        }
+
+        with self._lock:
+            self._version_line = self._verify_boot()
+
+        self._verify_firmware_version()
```

Comment:
```suggestion
        self._version_line = self._verify_boot()
        self._verify_firmware_version()
```

Don't need this lock. `_command` uses this lock too and it freezes the program.